### PR TITLE
Fix #1578 - Removes Alamofire

### LINKED ISSUE #1578: Consider droppping Alamofire dependency
`URLSession` should be sufficient for our needs.

We need to do research what Alamofire is used for and how easy will be to replace it with `URLSession` 

This is dev internal issue but I'm marking it as `QA/Yes` as some basic tests will need to be done to verify we converted alamofire calls properly

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary of Changes

Fixes: https://github.com/brave/brave-ios/issues/1578

Security Review for SSL Pinning: https://github.com/brave/internal/issues/665

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
1. Check if referrals are working.
2. Enable charles proxy and intercept ssl requests.
3. Verify that referrals are NOT working with charles proxy enabled.
4. Disable charles proxy and verify that referrals are working again.

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/internal/issues/new?assignees=&labels=security%2C+security-reviews&template=security-review.md&title=iOS:%20Security+Review+for) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: BraveShared/Analytics/UrpService.swift
Comment (kylehickinson) [inline BraveShared/Analytics/UrpService.swift]: Can this be moved into a separate file since they are generic/unrelated to URP?

File: BraveShared/Analytics/UrpService.swift
Comment (kylehickinson) [inline BraveShared/Analytics/UrpService.swift]: Same as above

File: BraveShared/Analytics/UrpService.swift
Comment (kylehickinson) [inline BraveShared/Analytics/UrpService.swift]: Same as above. `URLSessionExtensions.swift` maybe?

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Client.xcodeproj/project.pbxproj
Comment (Brandon-T) [inline Client.xcodeproj/project.pbxproj]: This is done by Xcode-11 automatically.. If I remove the new line, the next time I go to build settings.. it comes back..

File: BraveShared/CertificatePinning.swift
Comment (iccub) [inline BraveShared/CertificatePinning.swift:12]: Move this to `Client/Networking` folder

File: Push/PushClient.swift
Comment (iccub) [inline Push/PushClient.swift]: This file can be removed, rebase to latest development to see that it;s gone now

**Review by iccub — CHANGES_REQUESTED** [review]
Looks great and the code change isn't even that big, two things:
1. Rebase to latest development
2. Request a security review for it

File: BraveShared/CertificatePinning.swift
Comment (Brandon-T) [reply BraveShared/CertificatePinning.swift:12]: If I move it, how do I access it from BraveShared? That's where the UrpService is located.

Comment (iccub) [reply BraveShared/CertificatePinning.swift:12]: you are right, sorry

File: BraveShared/Analytics/URLSession+Requests.swift
Comment (iccub) [inline BraveShared/Analytics/URLSession+Requests.swift]: This line is too long, split it please, we try to aim for 125 line width max

File: BraveShared/CertificatePinning.swift
Comment (iccub) [inline BraveShared/CertificatePinning.swift]: `let error` can be ommited

File: Client.xcodeproj/project.pbxproj
Comment (iccub) [inline Client.xcodeproj/project.pbxproj]: `KIF` should be removed, we don't use it afaik

File: Client.xcodeproj/project.pbxproj
Comment (iccub) [inline Client.xcodeproj/project.pbxproj]: Remove `AdSupport`

**Review by iccub — CHANGES_REQUESTED** [review]

**Review by iccub — APPROVED** [review]
Looks good, don't merge until security review

**Comment by jumde** [comment]
Can you add a test to make sure that cert-pinning works?

File: BraveShared/CertificatePinning.swift
Comment (jumde) [inline BraveShared/CertificatePinning.swift:80]: Is this just for testing? If so can we gate this behind a flag to avoid using it in actual code?

File: BraveShared/Analytics/UrpService.swift
Comment (jumde) [inline BraveShared/Analytics/UrpService.swift:31]: no host-validation?

File: BraveShared/CertificatePinning.swift
Comment (jumde) [inline BraveShared/CertificatePinning.swift:137]: The certificates in the `Assets` directory are root certs for Global/DigiCert. Wouldn't it be better to fetch the certificate data of the root cert only? `SecTrustGetCertificateAtIndex` should get the root cert for index = max.

File: BraveShared/Analytics/UrpService.swift
Comment (Brandon-T) [reply BraveShared/Analytics/UrpService.swift:31]: By default, `options` is set to `[.default, .validateHost]` in the constructor.

File: BraveShared/CertificatePinning.swift
Comment (Brandon-T) [reply BraveShared/CertificatePinning.swift:80]: It is for testing. We shouldn't really be using `.allowSelfSigned` at all in any sort of production code. For that reason, it is NOT part of the default options. You'd have to explicitly set this flag in the options and I don't think anyone here would ever do that. Plus during a sec-review, if we see that option set, we should automatically reject the PR :)

It is hidden behind a flag already: `.allowSelfSigned` as part of the `OptionSet` and must be explicitly passed in if we want it.

File: BraveShared/CertificatePinning.swift
Comment (Brandon-T) [reply BraveShared/CertificatePinning.swift:137]: The code was done like this on purposed to validate all the certificates in the chain.
In my personal experience, I have never seen an app (until this one) that pins against the "root" certificate as it pretty much defeats the purpose of pinning unless you truly trust the CA (which I don't because I've seen them get compromised before).

Usually we pin against the Leaf certificate. As we go up the chain of certificates, security decreases. The leaf is the lowest in the chain, then intermediate, then root.

For this reason, I did not know why we were pinning the root but I decided to leave it how it originally was and just keep the validation for the chain (which should always happen). If any of the certificates are invalid, pinning should fail.

There's a nice article on pinning here: https://medium.com/@appmattus/android-security-ssl-pinning-1db8acb6621e

It's for Android, but it applies to every system. Specifically the section: **Which certificate should you pin against in the chain?**

I also did not know why we are doing binary pinning instead of public-key pinning but I left it how it originally was (hopefully the certs don't rotate).

IF we want, I can still change the validation to just the root but it would make sense to leave it how it is now so that we can validate the entire chain for future certificates as well.

Comment (jumde) [reply BraveShared/CertificatePinning.swift:137]: validating the complete chain is a good idea, but the leaf and intermediate certs change frequently and its hard to have a new release with every change. I think the change is good as is.

**Review by jumde — APPROVED** [review]

**Review by kylehickinson — APPROVED** [review]

**Review by iccub — APPROVED** [review]
Great work Brandon :)

**Pull request merged**
